driver-core/src/main/com/mongodb/internal/time/Timeout.java

```diff
@@ -94,7 +112,29 @@ default void awaitOn(final Condition condition) {
                 condition.awaitNanos(remaining(NANOSECONDS));
             }
         } catch (InterruptedException e) {
-            throw interruptAndCreateMongoInterruptedException(null, e);
+            throw interruptAndCreateMongoInterruptedException("Interrupted while " + action.get(), e);
+        }
+    }
+
+    /**
+     * {@linkplain CountDownLatch#await(long, TimeUnit) awaits} on the provided
+     * condition. Will {@linkplain CountDownLatch#await()} await} without a waiting
+     * time if this timeout is infinite. The {@link #hasExpired()} method is not
+     * checked by this method, and should be called outside of this method.
+     * @param latch the latch.
+     * @param action supplies the name of the action, for {@link MongoInterruptedException}
+     */
+    default void awaitOn(final CountDownLatch latch, final Supplier<String> action) {
+        try {
+            if (isInfinite()) {
+                latch.await();
+            } else {
+                // the timeout will track this remaining time
+                //noinspection ResultOfMethodCallIgnored
+                latch.await(this.remaining(NANOSECONDS), NANOSECONDS);
```

Comment:
The result of the latch needs to be inspected - as if it returns false it awaited the time limit but wasn't counted down.

Reply:
Previously, the 3 usages, all in BaseCluster, that were refactored into this method did not inspect the result. (Perhaps they should have inspected the result, but I could not readily see the error.)

Reply:
Agreed - nothing to do here.